File: verify.py
#!/usr/bin/env python3
"""Headless verification for The Algorithm.

Rebuilds `index.html` from source (runs build.py), then loads it in a headless
Chromium and checks the whole course actually works:

  * every lesson in content/ is present in the page,
  * every <img> resolves (no broken charts),
  * every quiz renders 4 options, shuffles, and grades on click,
  * every quiz exposes a working reset control,
  * the lightbox opens, browses the lesson's charts, zooms, and closes correctly,
  * each section's summary + final exam pages render, and the exam grades on submit,
  * a video link renders for each lesson that has a non-empty video.txt,
  * zero console/page JS errors.

Exit 0 on success, non-zero (with a list of problems) on any failure — so it
works both locally (`python verify.py`) and in CI.

Requires: pip install playwright && python -m playwright install chromium
"""
import subprocess, sys, pathlib

ROOT = pathlib.Path(__file__).resolve().parent


def main() -> int:
    # 1. Rebuild so we always verify the current source, never a stale artifact.
    if subprocess.run([sys.executable, str(ROOT / "build.py")]).returncode != 0:
        print("verify: build.py failed", file=sys.stderr)
        return 1

    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print("verify: playwright not installed — "
              "pip install playwright && python -m playwright install chromium",
              file=sys.stderr)
        return 2

    # Expected counts, derived from content/ (never hard-coded).
    lesson_files = list((ROOT / "content").glob("*/*/*/lesson.html"))
    expected_lessons = len(lesson_files)
    expected_videos = sum(
        1 for lh in lesson_files
        if (v := lh.parent / "video.txt").exists() and v.read_text(encoding="utf-8").strip()
    )
    expected_summaries = len(list((ROOT / "content").glob("*/summary.html")))
    expected_exams = len(list((ROOT / "content").glob("*/exam.js")))

    url = (ROOT / "index.html").resolve().as_uri()
    problems: list[str] = []

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_context(viewport={"width": 1400, "height": 900}).new_page()
        errs: list[str] = []
        page.on("pageerror", lambda e: errs.append(f"pageerror: {e}"))
        page.on("console", lambda m: errs.append(f"console.error: {m.text}")
                if m.type == "error" else None)
        page.goto(url, wait_until="networkidle")

        # Force every lesson visible + every chart image eager so lazy charts load.
        # Scope to `.fig img` — the injected chart images; ignore the lightbox <img>.
        page.evaluate(
            "() => { document.querySelectorAll('.lesson').forEach(s => s.classList.add('visible'));"
            "document.querySelectorAll('.fig img').forEach(i => { i.loading = 'eager'; i.src = i.src; }); }"
        )
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(400)

        page.on("dialog", lambda d: d.accept())

        def probe(expr, label, fallback):
            """Run a DOM check that must never mask a page error with a traceback.

            When app.js has died (e.g. a data block failed to parse) the widgets
            it renders don't exist, so these probes hit nulls. Report that as a
            problem — the real cause shows up in the JS-error list below."""
            try:
                return page.evaluate(expr)
            except Exception as exc:  # noqa: BLE001 - any page failure is a finding
                problems.append(f"{label} check could not run: {str(exc).splitlines()[0]}")
                return fallback

        # Review/exam pages are .lesson sections too — they carry data-kind and
        # must not be counted as lessons.
        ids = page.eval_on_selector_all(
            "section.lesson",
            "els => els.filter(e => e.id !== 'home' && !e.dataset.kind).map(e => e.id)")
        if len(ids) != expected_lessons:
            problems.append(f"lessons: page has {len(ids)}, content/ has {expected_lessons}")

        summaries = page.eval_on_selector_all('section.lesson[data-kind="review"]', "e => e.length")
        if summaries != expected_summaries:
            problems.append(f"summary pages: page has {summaries}, content/ has {expected_summaries}")
        exam_pages = page.eval_on_selector_all('section.lesson[data-kind="exam"]', "e => e.length")
        if exam_pages != expected_exams:
            problems.append(f"exam pages: page has {exam_pages}, content/ has {expected_exams}")

        imgs = page.eval_on_selector_all(
            ".fig img", "els => els.map(i => ({s: i.getAttribute('src'), nw: i.naturalWidth, c: i.complete}))")
        broken = [i["s"] for i in imgs if i["s"] and i["c"] and i["nw"] == 0]
        if broken:
            problems.append(f"{len(broken)} broken image(s): {broken[:5]}")

        videos = page.eval_on_selector_all(".lesson-video", "els => els.length")
        if videos != expected_videos:
            problems.append(f"video links: page has {videos}, content/ has {expected_videos}")

        quiz = probe(
            """() => { let qs = 0, bad = 0;
                 document.querySelectorAll('.quiz .q').forEach(q => {
                   const opts = [...q.querySelectorAll('.opt')]; qs++;
                   if (opts.length !== 4) { bad++; return; }
                   opts[0].click();
                   if (q.querySelectorAll('.opt.correct').length !== 1) bad++;
                 });
                 return { qs, bad }; }""",
            "quiz", {"qs": 0, "bad": 0}
        )
        if quiz["qs"] == 0:
            problems.append("no quiz questions rendered")
        if quiz["bad"]:
            problems.append(f"{quiz['bad']} quiz question(s) failed (not 4 options, or grading broke)")

        # Every quiz must offer a reset that actually clears the graded state.
        resets = probe(
            """() => { const qs = [...document.querySelectorAll('.quiz')];
                 const btns = qs.map(q => q.querySelector('.quiz-head .mini-btn'));
                 if (!qs.length || btns.some(b => !b)) return {missing: true};
                 btns[0].click();
                 const q = qs[0].querySelector('.q');
                 if (!q) return {missing: true};
                 return {missing: false, quizzes: qs.length,
                         cleared: q.querySelectorAll('.opt.correct, .opt.wrong').length === 0,
                         enabled: [...q.querySelectorAll('.opt')].every(o => !o.disabled)}; }""",
            "quiz reset", {"missing": True}
        )
        if resets.get("missing"):
            problems.append("at least one quiz has no reset button")
        elif not (resets["cleared"] and resets["enabled"]):
            problems.append("quiz reset did not clear the graded state")

        # Lightbox: opens on a chart, browses that lesson's set, zooms, and
        # closes on an outside click but NOT on a click on the image itself.
        # Uses REAL mouse input, not el.click() — while zoomed the stage holds a
        # pointer capture that retargets the click, and only real input sees it.
        def lb_state(what):
            return page.evaluate(
                """() => { const lb = document.getElementById('lightbox'),
                             big = lb.querySelector('img'),
                             t = k => (lb.querySelector(k) || {}).textContent || '';
                       return {open: lb.classList.contains('open'), src: big.src,
                               w: big.clientWidth, count: t('.lb-count'), zoom: t('.lb-zoom'),
                               lock: document.body.classList.contains('lb-lock')}; }""")
        try:
            first = page.query_selector(".lesson .fig img")
            if first is None:
                raise RuntimeError("no chart images to open")
            first.click(timeout=5000)
            page.wait_for_timeout(250)
            st = lb_state("open")
            if not st["open"]:
                problems.append("lightbox did not open on a chart click")
            if not st["lock"]:
                problems.append("lightbox did not lock background scrolling")
            total = st["count"].split("/")[-1].strip()
            if not total.isdigit() or int(total) < 2:
                problems.append(f"lightbox group looks wrong (counter {st['count']!r})")

            page.click('[data-lb="next"]', timeout=5000); page.wait_for_timeout(300)
            if lb_state("next")["src"] == st["src"]:
                problems.append("lightbox next did not change the image")
            page.click('[data-lb="prev"]', timeout=5000); page.wait_for_timeout(300)

            w0 = lb_state("fit")["w"]
            page.click('[data-lb="in"]', timeout=5000); page.wait_for_timeout(250)
            zs = lb_state("zoomed")
            if not (zs["w"] > w0 and zs["zoom"] == "125%"):
                problems.append(f"lightbox zoom-in failed ({w0} -> {zs['w']}, {zs['zoom']})")
            # A plain click on the image while zoomed must NOT close it.
            box = page.eval_on_selector(
                "#lightbox img",
                "e => { const r = e.getBoundingClientRect();"
                "return {x: r.x + r.width / 2, y: r.y + r.height / 2}; }")
            page.mouse.click(box["x"], box["y"]); page.wait_for_timeout(250)
            if not lb_state("after image click")["open"]:
                problems.append("lightbox closed when the image itself was clicked")
            page.click('[data-lb="reset"]', timeout=5000); page.wait_for_timeout(250)
            if lb_state("refit")["w"] != w0:
                problems.append("lightbox fit did not restore the original size")

            page.mouse.click(5, 5); page.wait_for_timeout(250)
            end = lb_state("closed")
            if end["open"]:
                problems.append("lightbox did not close on an outside click")
            if end["lock"]:
                problems.append("lightbox left the body scroll lock on after closing")
        except Exception as exc:  # noqa: BLE001 - any page failure is a finding
            problems.append(f"lightbox check could not run: {str(exc).splitlines()[0]}")
            page.evaluate("() => { document.getElementById('lightbox')"
                          ".classList.remove('open');"
                          "document.body.classList.remove('lb-lock'); }")

        # Answer a whole exam, submit, and confirm it grades to a real score.
        exam = probe(
            """() => { const out = [];
                 document.querySelectorAll('.exam').forEach(ex => {
                   const qs = [...ex.querySelectorAll('.q.eq')];
                   const bad = qs.filter(q => q.querySelectorAll('.opt').length !== 4).length;
                   qs.forEach(q => { const o = q.querySelector('.opt'); if (o) o.click(); });
                   const submit = ex.querySelector('.exam-actions .btn.primary');
                   if (submit) submit.click();
                   const score = ex.querySelector('.exam-score');
                   out.push({qs: qs.length, bad, submit: !!submit,
                             score: score ? score.textContent.trim() : null,
                             retake: !!ex.querySelector('.exam-actions .btn:not([style*="none"])')});
                 });
                 return out; }""",
            "exam", []
        )
        if len(exam) != expected_exams:
            problems.append(f"exams rendered: {len(exam)}, content/ has {expected_exams}")
        for i, e in enumerate(exam):
            if e["qs"] == 0:
                problems.append(f"exam {i}: no questions rendered")
            if e["bad"]:
                problems.append(f"exam {i}: {e['bad']} question(s) without 4 options")
            if not e["submit"]:
                problems.append(f"exam {i}: no submit button rendered")
            if not (e["score"] or "").endswith("%"):
                problems.append(f"exam {i}: submit did not produce a score (got {e['score']!r})")
            if not e["retake"]:
                problems.append(f"exam {i}: no retake control after submitting")

        if errs:
            problems.append(f"{len(errs)} JS error(s): {errs[:5]}")

        browser.close()

    if problems:
        print("VERIFY FAILED:", file=sys.stderr)
        for pr in problems:
            print(f"  - {pr}", file=sys.stderr)
        return 1

    imgs_ok = len([i for i in imgs if i["s"]])
    print(f"verify OK: {len(ids)} lessons, {imgs_ok} images, "
          f"{videos} video links, {quiz['qs']} quiz questions, "
          f"{summaries} summary page(s), {sum(e['qs'] for e in exam)} exam questions "
          f"across {len(exam)} exam(s), 0 JS errors")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: verify.mjs
// verify.mjs — headless smoke test for the built site (dist/).
// Mirrors the checks verify.py used to run: every lesson page exists and
// renders, quizzes grade and reset, the lightbox opens/zooms/navigates,
// exams submit with a score, themes switch, and there are zero console
// errors. Run: pnpm verify (after pnpm build).

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { createServer } from "node:http";
import { dirname, extname, join, normalize } from "node:path";
import { chromium } from "playwright";

const DIST = join(process.cwd(), "dist");
// Base path must match the one baked into the build (see astro.config.mjs):
// "/the-algorithm" for GitHub Pages, "" (root) for Coolify builds. Setting
// BASE_PATH at verify time keeps the checks aligned with the built site.
const BASE = (process.env.BASE_PATH ?? "/the-algorithm").replace(/\/+$/, "");
const MIME = {
  ".css": "text/css; charset=utf-8",
  ".html": "text/html; charset=utf-8",
  ".ico": "image/x-icon",
  ".js": "text/javascript; charset=utf-8",
  ".json": "application/json",
  ".png": "image/png",
  ".svg": "image/svg+xml",
  ".txt": "text/plain; charset=utf-8",
  ".webmanifest": "application/manifest+json",
};

/* ---------------- static server over dist/ (strips the base path) ------ */

function serve() {
  return createServer((req, res) => {
    let p = decodeURIComponent(new URL(req.url, "http://x").pathname);
    if (p.startsWith(BASE)) {
      p = p.slice(BASE.length);
    }
    if (p === "" || p === "/") {
      p = "/index.html";
    }
    let file = join(DIST, normalize(p));
    if (!file.startsWith(DIST)) {
      res.writeHead(403);
      res.end();
      return;
    }
    if (!existsSync(file) || statSync(file).isDirectory()) {
      file = join(DIST, p, "index.html");
    }
    if (!existsSync(file)) {
      file = join(DIST, "404.html");
    }
    res.writeHead(200, {
      "Content-Type": MIME[extname(file)] || "application/octet-stream",
    });
    res.end(readFileSync(file));
  });
}

/* ---------------- expectations derived from content/ -------------------- */

const SPLIT_RE = /[\\/]/;
const SEMI_BRACE_RE = /;\s*}/g;
const SEMI_END_RE = /;\s*$/;
const SLUG_RE = /data-slug="([^"]+)"/;
const TITLE_RE = /data-title="([^"]+)"/;
const ID_RE = /id\s*:\s*["']([^"']+)["']/;
const CHART_FILE_RE = /^(.+)-(\d{2})\.png$/;

const parts = (p) => p.split(SPLIT_RE).filter(Boolean);

function walk(dir) {
  const out = [];
  for (const name of readdirSync(dir)) {
    const full = join(dir, name);
    if (statSync(full).isDirectory()) {
      out.push(...walk(full));
    } else {
      out.push(full);
    }
  }
  return out;
}

const contentFiles = walk(join(process.cwd(), "content"));
const sectionIdByDir = new Map(); // content dir -> section id
for (const f of contentFiles.filter((x) => x.endsWith("section.js"))) {
  const m = readFileSync(f, "utf8").match(ID_RE);
  if (m) {
    sectionIdByDir.set(parts(f).at(-2), m[1]);
  }
}

const lessons = contentFiles.filter((f) => f.endsWith("lesson.html"));
const videoFiles = contentFiles.filter(
  (f) => f.endsWith("video.txt") && readFileSync(f, "utf8").trim().length > 0
);
const summaryFiles = contentFiles.filter((f) => f.endsWith("summary.html"));
const examFiles = contentFiles.filter((f) => f.endsWith("exam.js"));

const parseArr = (file) => {
  const src = readFileSync(file, "utf8")
    .trim()
    .replace(SEMI_BRACE_RE, "}")
    .replace(SEMI_END_RE, "");
  return new Function(`return (${src});`)();
};

// slug -> expected chart count (max NN seen in images/)
const chartCounts = new Map();
for (const f of walk(join(process.cwd(), "images"))) {
  const m = parts(f).at(-1).match(CHART_FILE_RE);
  if (!m) {
    continue;
  }
  const n = Number(m[2]);
  if (!chartCounts.has(m[1]) || chartCounts.get(m[1]) < n) {
    chartCounts.set(m[1], n);
  }
}

const lessonExpect = lessons.map((f) => {
  const p = parts(f);
  const html = readFileSync(f, "utf8");
  const id = p.at(-2);
  const [, slug] = html.match(SLUG_RE) || [];
  const [, title = ""] = html.match(TITLE_RE) || [];
  return {
    charts: slug ? chartCounts.get(slug) || 0 : 0,
    id,
    month: p.at(-2).slice(0, 2),
    quiz: parseArr(join(dirname(f), "quiz.js")),
    section: sectionIdByDir.get(p.at(-4)),
    title,
    video: videoFiles.some((v) => parts(v).at(-2) === id),
  };
});
const examExpect = examFiles.map((f) => ({
  questions: parseArr(f),
  section: sectionIdByDir.get(parts(f).at(-2)),
}));

/* ---------------- run --------------------------------------------------- */

let failures = 0;
const fail = (msg) => {
  failures += 1;
  console.error(`  ✗ ${msg}`);
};
const ok = (msg) => console.log(`  ✓ ${msg}`);

const server = serve();
await new Promise((r) => server.listen(0, r));
const origin = `http://127.0.0.1:${server.address().port}`;
const url = (p) => `${origin}${BASE}${p}`;

console.log("Routes…");
const routes = ["/", "/course"];
for (const f of summaryFiles) {
  routes.push(`/course/${sectionIdByDir.get(parts(f).at(-2))}/review`);
}
for (const f of examFiles) {
  routes.push(`/course/${sectionIdByDir.get(parts(f).at(-2))}/exam`);
}
for (const f of lessons) {
  const p = parts(f);
  routes.push(
    `/course/${sectionIdByDir.get(p.at(-3))}/${p.at(-2).slice(0, 2)}/${p.at(-2)}`
  );
}
for (const r of routes) {
  const res = await fetch(url(r));
  if (res.status !== 200) {
    fail(`route ${r} -> ${res.status}`);
  }
}
ok(`${routes.length} routes return 200`);

const browser = await chromium.launch();
const page = await browser.newPage();
const errors = [];
page.on("pageerror", (e) => errors.push(`pageerror: ${e.message}`));
page.on("console", (m) => {
  if (m.type() === "error") {
    errors.push(`console.error: ${m.text()}`);
  }
});

console.log("Landing…");
await page.goto(url("/"));
if (!(await page.locator(".home-hero h1").count())) {
  fail("landing hero missing");
}
const startHref = await page.locator("a.start").first().getAttribute("href");
if (startHref !== `${BASE}/course`) {
  fail(`start href = ${startHref}`);
}
ok("hero + start button");

console.log("Theme toggle…");
await page.goto(url("/course"));
await page.click(".theme-btn");
await page.locator(".theme-menu button", { hasText: "Dark" }).click();
const dark = await page.evaluate(() =>
  document.documentElement.getAttribute("data-theme")
);
if (dark !== "trading") {
  fail(`dark theme -> ${dark}`);
}
const stored = await page.evaluate(() => localStorage.getItem("ict-theme"));
if (stored !== "dark") {
  fail(`ict-theme stored = ${stored}`);
}
await page.reload();
const afterReload = await page.evaluate(() =>
  document.documentElement.getAttribute("data-theme")
);
if (afterReload !== "trading") {
  fail(`theme after reload -> ${afterReload}`);
}
await page.click(".theme-btn");
await page.locator(".theme-menu button", { hasText: "System" }).click();
await page.evaluate(() => localStorage.removeItem("ict-theme"));
ok("light/dark/system switching + persistence");

console.log("Course page…");
const mcardCount = await page.locator(".mcard").count();
if (mcardCount < 6) {
  fail(`month cards = ${mcardCount}`);
}
const secHeads = await page.locator(".sec-head").count();
if (secHeads !== 2) {
  fail(`sec-heads = ${secHeads}`);
}
if (!(await page.locator(".reset-panel").count())) {
  fail("reset panel missing");
}
ok(`month cards (${mcardCount}) + reset panel`);

console.log("Lessons…");
for (const exp of lessonExpect) {
  await page.goto(url(`/course/${exp.section}/${exp.month}/${exp.id}`));
  const h1 = (await page.locator(".lesson h1").textContent()) || "";
  if (h1.trim() !== exp.title) {
    fail(`${exp.id}: h1 "${h1.trim()}" != "${exp.title}"`);
  }
  const figs = await page.locator(".fig img").count();
  if (figs !== exp.charts) {
    fail(`${exp.id}: figs ${figs} != ${exp.charts}`);
  }
  const broken = await page
    .locator(".fig img")
    .evaluateAll((imgs) => imgs.filter((i) => i.naturalWidth === 0).length);
  if (broken > 0) {
    fail(`${exp.id}: ${broken} broken charts`);
  }
  const qs = await page.locator(".quiz .q").count();
  if (qs !== exp.quiz.length) {
    fail(`${exp.id}: questions ${qs} != ${exp.quiz.length}`);
  }
  for (let i = 0; i < qs; i += 1) {
    const opts = await page.locator(".quiz .q").nth(i).locator(".opt").count();
    if (opts !== 4) {
      fail(`${exp.id} q${i}: ${opts} options`);
    }
  }
  if (qs > 0) {
    await page.locator(".quiz").first().scrollIntoViewIfNeeded();
    await page
      .waitForFunction(
        () => {
          const el = document.querySelector(".quiz .opt");
          return !!el && Object.keys(el).some((k) => k.startsWith("__react"));
        },
        undefined,
        { timeout: 10_000 }
      )
      .catch(() => fail(`${exp.id}: quiz island never hydrated`));
    await page.locator(".quiz .q").first().locator(".opt").first().click();
    const correct = await page
      .locator(".quiz .q")
      .first()
      .locator(".opt.correct")
      .count();
    const expl = await page
      .locator(".quiz .q")
      .first()
      .locator(".expl.show")
      .count();
    if (correct !== 1) {
      fail(`${exp.id}: grading -> ${correct} correct`);
    }
    if (expl !== 1) {
      fail(`${exp.id}: explanation not shown`);
    }
    const resetEnabled = await page.locator(".mini-btn").isEnabled();
    if (!resetEnabled) {
      fail(`${exp.id}: reset disabled after answer`);
    }
    await page.locator(".mini-btn").click();
    const graded = await page.locator(".quiz .opt.correct").count();
    if (graded !== 0) {
      fail(`${exp.id}: reset did not clear grading`);
    }
  }
  const hasVideo = (await page.locator(".lesson-video").count()) > 0;
  if (hasVideo !== exp.video) {
    fail(`${exp.id}: video link mismatch`);
  }
  if (!(await page.locator(".notes-area").count())) {
    fail(`${exp.id}: notes box missing`);
  }
  if (!(await page.locator(".save-btn").count())) {
    fail(`${exp.id}: notes save button missing`);
  }
  if (!(await page.locator(".clear-btn").count())) {
    fail(`${exp.id}: notes clear button missing`);
  }
}
ok(
  `${lessonExpect.length} lessons: titles, charts, quizzes (4 opts, grade, reset), videos, notes`
);

console.log("Lightbox…");
const [chartLesson] = lessonExpect
  .filter((l) => l.charts >= 2)
  .sort((a, b) => b.charts - a.charts);
if (chartLesson) {
  await page.goto(
    url(`/course/${chartLesson.section}/${chartLesson.month}/${chartLesson.id}`)
  );
  await page
    .waitForFunction(() => document.documentElement.dataset.lbReady === "true")
    .catch(() => fail("lightbox island never hydrated"));
  await page.locator(".fig img").first().click();
  await page.waitForTimeout(300);
  if (!(await page.locator(".yarl__portal_open").count())) {
    fail("lightbox did not open");
  }
  if (
    !(await page.evaluate(() =>
      document.body.classList.contains("yarl__no_scroll")
    ))
  ) {
    fail("body not locked");
  }
  // Verify navigation buttons exist and clicking next keeps the lightbox open.
  if (!(await page.locator(".yarl__navigation_prev").count())) {
    fail("prev button missing");
  }
  if (!(await page.locator(".yarl__navigation_next").count())) {
    fail("next button missing");
  }
  await page.click(".yarl__navigation_next");
  await page.waitForTimeout(400);
  if (!(await page.locator(".yarl__portal_open").count())) {
    fail("lightbox closed after next click");
  }
  // Check that the zoom-in toolbar button exists (zoom plugin is active).
  const zoomInBtn = page.locator(".yarl__toolbar button[aria-label='Zoom in']");
  if (!(await zoomInBtn.count())) {
    fail("zoom-in button missing");
  }
  // Close with Escape.
  await page.keyboard.press("Escape");
  await page.waitForTimeout(400);
  if (await page.locator(".yarl__portal_open").count()) {
    fail("Escape did not close");
  }
  // Re-open via click and close again.
  await page.locator(".fig img").first().click();
  await page.waitForTimeout(300);
  if (!(await page.locator(".yarl__portal_open").count())) {
    fail("lightbox did not re-open");
  }
  await page.keyboard.press("Escape");
  await page.waitForTimeout(200);
  if (await page.locator(".yarl__portal_open").count()) {
    fail("Escape did not close on re-open");
  }
  ok("open / navigation / zoom button / close / re-open behaviors");
} else {
  fail("no lesson with >= 2 charts found");
}

console.log("Exams…");
for (const exp of examExpect) {
  await page.goto(url(`/course/${exp.section}/exam`));
  const qs = await page.locator(".exam .q").count();
  if (qs !== exp.questions.length) {
    fail(`${exp.section}: questions ${qs} != ${exp.questions.length}`);
  }
  await page.locator(".exam").first().scrollIntoViewIfNeeded();
  await page
    .waitForFunction(
      () => {
        const el = document.querySelector(".exam .opt");
        return !!el && Object.keys(el).some((k) => k.startsWith("__react"));
      },
      undefined,
      { timeout: 10_000 }
    )
    .catch(() => fail(`${exp.section}: exam island never hydrated`));
  for (let i = 0; i < qs; i += 1) {
    const opts = await page.locator(".exam .q").nth(i).locator(".opt").count();
    if (opts !== 4) {
      fail(`${exp.section} q${i}: ${opts} options`);
    }
  }
  const badge = await page.locator(".exam-best").textContent();
  if (!badge.includes("Not attempted yet")) {
    fail(`${exp.section}: badge = "${badge}"`);
  }
  const submitDisabled0 = await page
    .locator(".exam-actions .primary")
    .isDisabled();
  if (!submitDisabled0) {
    fail(`${exp.section}: submit enabled with 0 answers`);
  }
  for (let i = 0; i < qs; i += 1) {
    const q = exp.questions[i];
    const texts = await page
      .locator(".exam .q")
      .nth(i)
      .locator(".opt")
      .allTextContents();
    const k = texts.findIndex((t) => t.trim() === q.o[q.a].trim());
    if (k === -1) {
      fail(`${exp.section} q${i}: correct option not found`);
      continue;
    }
    await page.locator(".exam .q").nth(i).locator(".opt").nth(k).click();
  }
  await page.click(".exam-actions .primary");
  await page.waitForTimeout(100);
  const score = (await page.locator(".exam-score").textContent()) || "";
  if (!/^\d+%$/.test(score.trim())) {
    fail(`${exp.section}: score = "${score}"`);
  }
  if (score.trim() !== "100%") {
    fail(`${exp.section}: expected 100%, got ${score.trim()}`);
  }
  if (!(await page.locator(".exam-score.pass").count())) {
    fail(`${exp.section}: not marked pass`);
  }
  if (
    !(await page.locator(".exam-actions .btn", { hasText: "Retake" }).count())
  ) {
    fail(`${exp.section}: retake missing`);
  }
  await page.locator(".exam-actions .btn", { hasText: "Retake" }).click();
  await page.waitForTimeout(50);
  const picked = await page.locator(".exam .opt.picked").count();
  if (picked !== 0) {
    fail(`${exp.section}: picks not cleared on retake`);
  }
  // After a retake nothing is answered, so submit is disabled again (0/m).
  const submitDisabled = await page
    .locator(".exam-actions .primary")
    .isDisabled();
  if (!submitDisabled) {
    fail(`${exp.section}: submit should be disabled after retake`);
  }
}
ok(
  `${examExpect.length} exams: 4 opts, full-answer submit -> 100% pass, retake`
);

console.log("Summary vs exam counts…");
// A section summary states its exam's question count in prose, and nothing
// else couples the two — they have drifted before. Check the stated number
// against the exam that actually renders.
const SUMMARY_EXAM_COUNT_RE = /Final Exam.{0,80}?(\d+)\s+questions?/s;
for (const f of summaryFiles) {
  const sec = sectionIdByDir.get(parts(f).at(-2));
  const m = readFileSync(f, "utf8").match(SUMMARY_EXAM_COUNT_RE);
  if (!m) {
    continue; // summary states no count — fine
  }
  const stated = Number(m[1]);
  const exam = examExpect.find((e) => e.section === sec);
  if (!exam) {
    fail(
      `${sec}: summary.html states ${stated} question(s) but no exam.js exists`
    );
    continue;
  }
  if (stated !== exam.questions.length) {
    fail(
      `${sec}: summary.html says ${stated}, exam.js has ${exam.questions.length}`
    );
  }
}
ok("summary-stated exam question counts match the rendered exams");

console.log("Review pages…");
for (const f of summaryFiles) {
  const sec = sectionIdByDir.get(parts(f).at(-2));
  await page.goto(url(`/course/${sec}/review`));
  const h1 = (await page.locator(".lesson h1").textContent()) || "";
  if (h1.trim() !== "Section Summary") {
    fail(`${sec}: review h1 = "${h1.trim()}"`);
  }
  const bodyLen = await page
    .locator(".lesson")
    .first()
    .evaluate((el) => el.textContent.length);
  if (bodyLen < 500) {
    fail(`${sec}: review body too short (${bodyLen})`);
  }
}
ok(`${summaryFiles.length} review pages render`);

console.log("Console…");
if (errors.length) {
  for (const e of errors.slice(0, 10)) {
    fail(e);
  }
} else {
  ok("zero page/console errors");
}

await browser.close();
server.close();
console.log(
  failures === 0 ? "\nAll checks passed." : `\n${failures} check(s) FAILED.`
);
process.exit(failures === 0 ? 0 : 1);
